onnxruntime/core/mlas/lib/jblas_gemm.cpp

```diff
@@ -0,0 +1,525 @@
+                    } else if (NTile == tAVX_VNNI::NTILE && _cd->AVX_VNNI()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAVX_VNNI>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+                if (CType == uint32_t(gemm::CompType::COMP_INT8_SS_INT32)) {
+                    if (NTile == tAMX_INT8_SS::NTILE && _cd->AMX_INT8()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAMX_INT8_SS>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+            }
+            delete ptr;
+        }
+    }
+    return size;
+}
+
+template <typename T>
+static size_t
+JblasQ4BuSize(int block_size, size_t N, size_t K, bool isAsym)
+{
+    static T launcher;
+    auto stor = launcher.mProB.createStorage(
+        N, K, block_size, JBLAS_DTYPE::S4_CLIP, JBLAS_DTYPE::F32, JBLAS_DTYPE::BF16, isAsym
+    );
+    // TODO(Yu) support more scale dtype
+    return stor.mSize;
+}
+
+size_t
+JblasQ4GemmPackBSize(size_t N, size_t K, size_t BlkSize, bool isAsym, MLAS_SQNBIT_COMPUTE_TYPE CompType)
+{
+    GetCPUDevice();
+    if (K % BlkSize != 0) {
+        return 0;
+    }
+    // from low precision to high precision
+    switch (CompType) {
+        case CompInt8:
+            if (_cd->AMX_INT8() && BlkSize % tAMX_INT8_SS::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAMX_INT8_SS>>(int(BlkSize), N, K, isAsym);
+            }
+            if (_cd->AVX512_VNNI() && BlkSize % tAVX512_VNNI::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX512_VNNI>>(int(BlkSize), N, K, isAsym);
+            }
+            if (_cd->AVX_VNNI() && BlkSize % tAVX_VNNI::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX_VNNI>>(int(BlkSize), N, K, isAsym);
+            }
+        case CompBf16:
+        case CompFp16:
+        case CompFp32:
+        case CompUndef:
+            if (_cd->AVX512F() && BlkSize % tAVX512F::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX512F>>(int(BlkSize), N, K, isAsym);
+            }
+            if (_cd->AVX2() && BlkSize % tAVX2::KTILE == 0) {
+                return JblasQ4BuSize<tLauncher_Int8_S4_F32F32<tAVX2>>(int(BlkSize), N, K, isAsym);
+            }
+            break;
+        default:
+            return 0;
+    }
+    return 0;
+}
+
+template <typename T>
+static void
+JblasQ4GemmPackBImpl(
+    void* PackedBuf,
+    int BlkSize,
```

Comment:
Are there any circumstances in which `BlkSize`, `N`, `K`, or `ldb` could take negative values, thus justifying the use of `int` instead of `size_t`? If not, it may be more suitable to declare these variables as `size_t`. This would also remove the need for casting in the `JblasQ4GemmPackB` function that calls this one.

Reply:
I've changed `N`, `K`, and `BlkSize` to `size_t` untill calling the Jblas' interface functions. BlkSize==-1 is reserved for per-channel quantization. But this is not enabled in MLAS' integration.